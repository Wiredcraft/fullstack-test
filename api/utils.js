const sumVotes = votes => {
  if (votes) {
    return votes.reduce((acc, cur) => {
      return acc + cur.val;
    }, 0);
  } else {
    return 0;
  }
};

module.exports = {
  sumVotes
};
