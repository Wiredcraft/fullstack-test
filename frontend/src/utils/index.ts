import * as dayjs from 'dayjs'
import relativeTime from 'dayjs/plugin/relativeTime';
dayjs.extend(relativeTime);

export function formatTime(date: Date) {
  return dayjs(date).fromNow()
}