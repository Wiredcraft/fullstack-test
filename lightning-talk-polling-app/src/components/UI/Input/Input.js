import React from 'react';
import AuxiliaryComponent from '../../../hoc/AuxiliaryComponent';

const input = (props) => (
    <AuxiliaryComponent>
        <label className="grey-text pt-4">{props.label}</label>
        <input
            type={props.type}
            className="form-control"
            placeholder={props.placeholder}
            maxLength={props.maxLength || 200}
            onChange={props.onChange}
        />
    </AuxiliaryComponent>
);

export default input;