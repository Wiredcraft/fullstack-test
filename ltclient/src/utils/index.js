import fetch from 'isomorphic-fetch';

let MIN = 60 * 1000;
let HOUR = MIN * 60;
let DAY = HOUR * 24;
let DAY3 = DAY * 3;

// return a friendly time
// '2016-08-24T06:50:06.018Z' => '3 hours ago'
function when(t) {
  if (typeof t === 'string') t = new Date(t);
  t  = new Date() - t;
  if (t > DAY3) {
    return 'on ' + t.toLocaleDateString();
  } else if (t > DAY) {
    return Math.floor(t / DAY) + ' days ago';
  } else if ( t > HOUR) {
    return Math.floor(t / HOUR) + ' hours ago';
  } else if ( t > MIN) {
    return Math.floor(t / MIN) + ' minutes ago';
  } else {
    return 'just now';
  }
}

function commonFetch(url, method, data) {
  let opt = {};
  opt.method = method || 'GET';
  opt.headers = {
    'Accept': 'application/json',
    'Content-Type': 'application/json'
  };

  if (data) {
    opt.body = JSON.stringify(data);
  }

  return fetch(url, opt);
}



export { when, commonFetch };