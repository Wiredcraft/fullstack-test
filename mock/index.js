const express = require('express')
const faker = require('faker')

const app = express()

app.set('port', process.env.PORT || 3001)

app.use(express.json())

const makeTalk = () => {
  return {
    author: faker.internet.userName(),
    title: faker.hacker.phrase(),
    description: faker.lorem.paragraphs(),
    id: faker.random.uuid(),
    created: new Date(faker.date.past()).getTime(),
    votes: faker.random.number(1000),
    voted: faker.random.boolean(),
    isPublic: faker.random.boolean(),
    publishDate: new Date(faker.date.past()).getTime(),
  }
}

app.get('/api/talks', (req, res) => {
  const talkList = []
  let i = Math.random() * 20
  while (i > 0) {
    talkList.push(makeTalk())
    i = i - 1
  }
  res.json(talkList)
})

app.post('/api/talks', (req, res) => {
  if (Math.random() > 0.5) {
    res.status(201).json(req.body)
  } else {
    res.status(400).json({
      violations: {
        title: 'Title exists.',
      },
    })
  }
})

app.post('/api/talks/:id/vote', (req, res) => {
  if (Math.random() > 0.1) {
    res.status(204).send()
  } else {
    res.status(500).send()
  }
})

app.listen(app.get('port'), () => {
  console.log(`Find the server at: http://localhost:${app.get('port')}/`) // eslint-disable-line no-console
})
