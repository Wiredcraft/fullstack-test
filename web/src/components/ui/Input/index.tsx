import React, { FC, InputHTMLAttributes } from "react";
import styled from "styled-components";

const Container = styled.input`
  color: ${({ theme: { colors } }) => colors.text.base};
  width: 100%;
  height: 32px;
  background: transparent;
  outline: none;
  border: none;
  box-shadow: 0 1px 0 ${({ theme: { colors } }) => colors.border.base};
  transition: box-shadow, 0.3s ease;
  &:focus {
    box-shadow: 0 3px 0 ${({ theme: { colors } }) => colors.border.light};
  }
`;

type Props = InputHTMLAttributes<HTMLInputElement>;

const Input: FC<Props> = ({ ...props }) => {
  return <Container {...props} />;
};

export default Input;
