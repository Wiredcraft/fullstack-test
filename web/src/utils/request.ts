import axios from "axios";
import { STORE_USER_INFO } from "../constants/auth";

const BASE_URL = "http://localhost:8383";
const instance = axios.create({
  baseURL: BASE_URL,
  timeout: 5000
});

instance.interceptors.request.use(config => {
  const { headers = {} } = config;
  const userString = localStorage.getItem(STORE_USER_INFO);
  if (userString) {
    try {
      const { token } = JSON.parse(userString);
      if (token) {
        headers.Authorization = `Bearer ${token}`;
      }
      // eslint-disable-next-line no-empty
    } catch (e) {}
  }
  return config;
});

instance.interceptors.response.use(
  response => {
    if (response.data.success) {
      return response.data.result;
    }
    throw response.data.reason;
  },
  error => Promise.reject(error)
);

export default instance;
