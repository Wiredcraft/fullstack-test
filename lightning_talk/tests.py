from django.contrib.auth.models import User
from rest_framework import status
from rest_framework.test import APITestCase
from rest_framework_jwt.utils import jwt_payload_handler, jwt_encode_handler
from .models import Post, Upvote


class PostTest(APITestCase):
    def setUp(self):
        self.user = User.objects.create_user(
            username='casca', password='pwd@1234')
        payload = jwt_payload_handler(self.user)
        self.token = jwt_encode_handler(payload)

    def test_create_post(self):
        """
        Create a new post object.
        """
        url = '/lightning_talk/api/posts/'
        data = {'title': 'yo', 'description': 'hei'}
        response = self.client.post(
            url, data, format='json', HTTP_AUTHORIZATION='JWT ' + self.token)
        self.assertEqual(response.status_code, status.HTTP_201_CREATED)
        self.assertEqual(Post.objects.count(), 1)
        self.assertEqual(Post.objects.get().description, 'hei')

    def test_update_post(self):
        """
        Update a post object
        """
        post = Post.objects.create(
            title='yo', description='hei', user=self.user)

        url = '/lightning_talk/api/posts/%s/' % post.pk
        data = {'description': 'hei ho'}
        response = self.client.patch(
            url, data, format='json', HTTP_AUTHORIZATION='JWT ' + self.token)
        self.assertEqual(response.status_code, status.HTTP_200_OK)
        self.assertEqual(Post.objects.get().description, 'hei ho')

    def test_delete_post(self):
        """
        Delete a post object
        """
        post = Post.objects.create(
            title='yo', description='hei', user=self.user)

        url = '/lightning_talk/api/posts/%s/' % post.pk
        response = self.client.delete(
            url, HTTP_AUTHORIZATION='JWT ' + self.token)
        self.assertEqual(response.status_code, status.HTTP_204_NO_CONTENT)
        self.assertEqual(Post.objects.count(), 0)

    def test_upvote_post(self):
        """
        Upvote post
        """
        post = Post.objects.create(
            title='yo', description='hei', user=self.user)

        url = '/lightning_talk/api/posts/%s/upvote/' % post.pk
        response = self.client.put(url, HTTP_AUTHORIZATION='JWT ' + self.token)
        self.assertEqual(response.status_code, status.HTTP_200_OK)
        self.assertEqual(Post.objects.get().upvotes.count(), 1)

    def test_cancle_upvote_of_post(self):
        """
        Cancle upvote post
        """
        post = Post.objects.create(
            title='yo', description='hei', user=self.user)
        upvote = Upvote.objects.create(user=self.user, post=post)
        post.upvotes.add(upvote)

        url = '/lightning_talk/api/posts/%s/upvote/' % post.pk
        response = self.client.delete(
            url, HTTP_AUTHORIZATION='JWT ' + self.token)
        self.assertEqual(response.status_code, status.HTTP_200_OK)
        self.assertEqual(Post.objects.get().upvotes.count(), 0)
