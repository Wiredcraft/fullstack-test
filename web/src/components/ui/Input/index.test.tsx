import React from "react";
import "jest-styled-components";
import { render } from "../../../../tests/utils";
import "../../../../tests/mock";
import Input from ".";

describe("test <Input />", () => {
  it("should match snapshot", () => {
    const { container } = render(<Input />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
