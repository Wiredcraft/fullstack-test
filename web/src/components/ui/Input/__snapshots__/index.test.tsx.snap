// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test <Input /> should match snapshot 1`] = `
.c0 {
  color: #17171D;
  width: 100%;
  height: 32px;
  background: transparent;
  outline: none;
  border: none;
  box-shadow: 0 1px 0 #636477;
  -webkit-transition: box-shadow,0.3s ease;
  transition: box-shadow,0.3s ease;
}

.c0:focus {
  box-shadow: 0 3px 0 #3B66F5;
}

<input
  class="c0"
/>
`;
